### PR TITLE
feat(stream): log the event when stream copy failed

### PR DESCRIPTION
**Pull request by 7sunarni**
**- What I met**
I met a problem when I ran KinD failed in my machine,  and I found that KinD can not write it's config to docker container from KinD log. 

After I checked the KinD code, I realized that when ran `echo "hello" | docker exec -i $CONTAINER cat /dev/stdin` will not get any output in my machine (works well in other machine).  So, there must be a problem in somewhere.

Then I spent lots of time on docker cli, docker daemon, containerd, containerd-shim to find where the input or output data lost, because there is not failed log. 

I hope this commit will help people save their time if they have same problem.

**- What I did**
add failed log when copy from stdin to iop pipe stdin

**- How I did it**

**- How to verify it**

**- Description for the changelog**

```markdown changelog

```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by 7sunarni** [comment]
@cpuguy83 hello~ can you take a look here and take a review?

File: container/stream/streams.go [hunk 1]
```diff
@@ -119,28 +119,31 @@ func (c *Config) CopyToPipe(iop *cio.DirectIO) {
 	ctx := context.TODO()
 
 	c.dio = iop
-	copyFunc := func(w io.Writer, r io.ReadCloser) {
+	copyFunc := func(name string, w io.Writer, r io.ReadCloser) {
 		c.wg.Add(1)
 		go func() {
 			if _, err := pools.Copy(w, r); err != nil {
-				log.G(ctx).Errorf("stream copy error: %v", err)
+				log.G(ctx).Errorf("stream copy %s error: %v", name, err)
```
Comment (thaJeztah) [inline container/stream/streams.go:126]: Not critical, but as we're updating this, we could consider using structured logs / fields for these, e.g.;


```go
log.G(ctx).WithFields(log.Fields{"error": err, "stream": name}).Error("copy stream")
```


☝️ open to suggestions for the message itself; not sure if "stream" is the right term to use for all of these

File: container/stream/streams.go [hunk 2]
```diff
@@ -119,28 +119,31 @@ func (c *Config) CopyToPipe(iop *cio.DirectIO) {
...
+		copyFunc("stderr", c.Stderr(), iop.Stderr)
 	}
 
 	if stdin := c.Stdin(); stdin != nil {
 		if iop.Stdin != nil {
 			go func() {
-				pools.Copy(iop.Stdin, stdin)
+				_, err := pools.Copy(iop.Stdin, stdin)
+				if err != nil {
+					log.G(ctx).Errorf("stream copy stdin error: %v", err)
```
Comment (thaJeztah) [inline container/stream/streams.go:145]: Same here;

```go
log.G(ctx).WithFields(log.Fields{"error": err, "stream": "stdin"}).Error("copy stream")
```

File: container/stream/streams.go [hunk 3]
```diff
@@ -119,28 +119,31 @@ func (c *Config) CopyToPipe(iop *cio.DirectIO) {
...
+				}
 				if err := iop.Stdin.Close(); err != nil {
 					log.G(ctx).Warnf("failed to close stdin: %v", err)
```
Comment (thaJeztah) [inline container/stream/streams.go:148]: (if we decide to update - message itself also open for suggestions, like the other ones);

```go
log.G(ctx).WithFields(log.Fields{"error": err, "stream": "stdin"}).Warn("closing stream")
```

**Review by thaJeztah — COMMENTED** [review]
Thanks!

At a quick glance, this seems fine to me; I left some suggestions, and can wait for others to have a look in case I overlooked any reason not to log these.


(FWIW; some maintainers are currently on vacation, so some delay may happen w.r.t. reviews / feedback)

**Comment by 7sunarni** [comment]
Yes, you are right. I just adjust the log to make likes other log output in the project.

File: container/stream/streams.go [hunk 4]
```diff
@@ -119,30 +119,35 @@ func (c *Config) CopyToPipe(iop *cio.DirectIO) {
 	ctx := context.TODO()
 
 	c.dio = iop
-	copyFunc := func(w io.Writer, r io.ReadCloser) {
+	copyFunc := func(name string, w io.Writer, r io.ReadCloser) {
 		c.wg.Add(1)
 		go func() {
 			if _, err := pools.Copy(w, r); err != nil {
-				log.G(ctx).Errorf("stream copy error: %v", err)
+				log.G(ctx).WithFields(log.Fields{"stream": name, "error": err}).Error(err)
```
Comment (thaJeztah) [inline container/stream/streams.go:126]: Looks like this now logs the error twice (once as log message, and once as `error` attribute); we can probably just keep the original log message;

```go
log.G(ctx).WithFields(log.Fields{"stream": name, "error": err}).Error("stream copy error")
```


That would show as something like;

```console
ERROR[2024-08-15T09:45:03.587172878Z] stream copy error                 error="something bad happened" stream="stdout"
```

File: container/stream/streams.go [hunk 5]
```diff
@@ -119,30 +119,35 @@ func (c *Config) CopyToPipe(iop *cio.DirectIO) {
...
+		copyFunc("stderr", c.Stderr(), iop.Stderr)
 	}
 
 	if stdin := c.Stdin(); stdin != nil {
 		if iop.Stdin != nil {
 			go func() {
-				pools.Copy(iop.Stdin, stdin)
+				_, err := pools.Copy(iop.Stdin, stdin)
+				if err != nil {
+					log.G(ctx).WithFields(log.Fields{"stream": "stdin", "error": err}).Error(err)
```
Comment (thaJeztah) [inline container/stream/streams.go:147]: Same for this one;

```go
log.G(ctx).WithFields(log.Fields{"stream": name, "error": err}).Error("stream copy error")
```

**Comment by 7sunarni** [comment]
Fine. I adjusted the error message.

**Review by vvoland — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

sorry for the delay, thought I already submitted 🙈

**Pull request merged**